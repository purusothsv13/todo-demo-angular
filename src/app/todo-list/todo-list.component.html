<ul>
  <div>Last updated: {{ date | date : "MM/dd/yyy" }}</div>
  <div>Search: <input [(ngModel)]="searchValue" type="text" /></div>
  <div
    *ngFor="
      let todoItem of todoItems | filterTasks : 'description' : searchValue;
      let i = index
    "
    style="display: flex; gap: 1rem; margin-bottom: 1rem"
  >
    <app-todo-list-item
      data-testid="todo-list-item"
      #listItem
      [item]="todoItem"
      [(progress)]="progress"
    >
      {{ i + 1 }}
    </app-todo-list-item>
    <button
      data-testid="todo-list-item-delete-btn"
      (click)="onDelete(todoItem)"
    >
      Delete
    </button>
    <button #viewBtn (click)="onView(todoItem)">View</button>
  </div>

  <app-create-todo-list-item (createTask)="onCreate($event)" />
</ul>
