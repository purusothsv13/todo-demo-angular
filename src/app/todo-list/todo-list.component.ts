import { CommonModule } from '@angular/common';
import {
  ChangeDetectionStrategy,
  ChangeDetectorRef,
  Component,
  ViewChild,
} from '@angular/core';
import { FormsModule } from '@angular/forms';
import { CreateTodoListItemComponent } from './create-todo-list-item/create-todo-list-item.component';
import { TodoListItemComponent } from './todo-list-item/todo-list-item.component';
import { TodoItem } from '../app.types';
import { FilterTasksPipe } from '../filter-tasks.pipe';
import { TodoDataService } from '../shared/todo-data.service';
import { ActivatedRoute, Router } from '@angular/router';
import {
  Observable,
  Subject,
  combineLatest,
  concat,
  filter,
  forkJoin,
  fromEvent,
  iif,
  interval,
  map,
  merge,
  of,
  partition,
  range,
  tap,
} from 'rxjs';

@Component({
  selector: 'app-todo-list',
  standalone: true,
  imports: [
    TodoListItemComponent,
    CreateTodoListItemComponent,
    FilterTasksPipe,
    FormsModule,
    CommonModule,
  ],
  templateUrl: './todo-list.component.html',
  styleUrl: './todo-list.component.scss',
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class TodoListComponent {
  @ViewChild(TodoListItemComponent) listItem!: TodoListItemComponent;
  @ViewChild('viewBtn') viewBtn!: HTMLButtonElement;

  date = new Date();

  progress = 0.7;
  searchValue = '';

  todoItems: TodoItem[] = [];

  constructor(
    private readonly todoDataService: TodoDataService,
    private readonly router: Router,
    private readonly route: ActivatedRoute,
    private readonly ref: ChangeDetectorRef
  ) {}

  ngOnInit() {
    const range$ = range(1, 10);

    const result = partition(range$, (n, i) => {
      return n % 2 === 0;
    });
    result[0].subscribe((val) => console.log('Even', val));

    concat(...result).subscribe((val) => {
      console.log(val);
    });

    merge(...result).subscribe((val) => {
      console.log('merge', val);
    });

    this.todoDataService.setSubject(1);
    this.todoDataService.setSubject(2);
    this.todoDataService.setSubject(3);
    this.todoDataService.setSubject(4);
    this.todoDataService.getSubject().subscribe((val) => {
      console.log('Component1', val);
    });

    this.todoDataService
      .getTodoItems$()
      .pipe(
        map((items) => {
          return items.filter((i) => !!i.title);
        })
      )
      .subscribe((items) => {
        this.todoItems = items;
        this.ref.markForCheck();
      });
  }

  onCreate(newTask: string): void {
    const todoItem = {
      id: Math.round(Math.random() * 10000),
      completed: false,
      title: newTask,
      userId: 1,
    };

    this.todoDataService.addTodoItem(todoItem).subscribe();
  }

  onView(item: TodoItem) {
    this.router.navigate([item.id], { relativeTo: this.route });
  }

  onDelete(item: TodoItem) {
    this.todoItems.splice(this.todoItems.indexOf(item), 1);
  }
}
