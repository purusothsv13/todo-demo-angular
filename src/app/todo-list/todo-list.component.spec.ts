import { ComponentFixture, TestBed } from '@angular/core/testing';

import { TodoListComponent } from './todo-list.component';
import { TodoDataService } from '../shared/todo-data.service';
import { ActivatedRoute, Router } from '@angular/router';
import { of } from 'rxjs';
import { By } from '@angular/platform-browser';
import { CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';

describe('TodoListComponent', () => {
  let component: TodoListComponent;
  let fixture: ComponentFixture<TodoListComponent>;
  let router: any;

  let itemsMock = [
    {
      title: 'Task1',
      completed: false,
    },
    {
      title: 'Task2',
      completed: true,
    },
  ];

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      imports: [TodoListComponent],
      providers: [
        {
          provide: TodoDataService,
          useValue: {
            getTodoItems$: jest.fn().mockReturnValue(of(itemsMock)),

            addTodoItem: jest.fn(),
          },
        },
        {
          provide: Router,
          useValue: {
            navigate: jest.fn(),
          },
        },
        {
          provide: ActivatedRoute,
          useValue: 'ACTIVATED_ROUTE',
        },
      ],
      schemas: [CUSTOM_ELEMENTS_SCHEMA],
    }).compileComponents();

    fixture = TestBed.createComponent(TodoListComponent);
    component = fixture.componentInstance;
    router = TestBed.inject(Router);
  });

  it('should create', () => {
    fixture.detectChanges();
    expect(component['todoDataService'].getTodoItems$).toHaveBeenCalled();
  });

  it('should assign items', () => {
    fixture.detectChanges();
    expect(component).toBeTruthy();
    expect(component.todoItems).toBeTruthy();
    expect(component.progress).toMatchSnapshot('progress');
    expect(component.todoItems).toMatchSnapshot('todoItems');
  });

  it('should filter out items with empty title', () => {
    component['todoDataService'].getTodoItems$ = jest.fn().mockReturnValue(
      of([
        {
          title: '',
        },
        {
          title: 'Task1',
        },
      ])
    );

    fixture.detectChanges();
    expect(component).toBeTruthy();
    expect(component.todoItems).toBeTruthy();
    expect(component.todoItems).toMatchSnapshot('todoItems');
  });

  it('should navigate to details page', () => {
    fixture.detectChanges();

    const item: any = {
      id: 'ITEM_ID_1',
    };

    component.onView(item);

    expect(router.navigate.mock.calls[0]).toMatchSnapshot();
  });

  it('should not navigate to details page', () => {
    fixture.detectChanges();

    component.onView(null);

    expect(router.navigate).not.toHaveBeenCalled();
  });

  describe('template', () => {
    it('should render todo list item', () => {
      fixture.detectChanges();

      const todoListItemElements = fixture.debugElement.queryAll(
        By.css('[data-testid="todo-list-item"]')
      );

      expect(todoListItemElements.length).toEqual(2);
    });

    it('should render todo list item', () => {
      const deleteSpy = jest.spyOn(component, 'onDelete');
      fixture.detectChanges();

      const todoListItemDelBtnElement = fixture.debugElement.query(
        By.css('[data-testid="todo-list-item-delete-btn"]')
      );

      todoListItemDelBtnElement.triggerEventHandler('click');

      expect(deleteSpy.mock.calls).toMatchSnapshot();
    });
  });
});
