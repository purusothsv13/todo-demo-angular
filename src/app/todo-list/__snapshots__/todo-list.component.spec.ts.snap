// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoListComponent should assign items: progress 1`] = `0.7`;

exports[`TodoListComponent should assign items: todoItems 1`] = `
[
  {
    "completed": false,
    "title": "Task1",
  },
  {
    "completed": true,
    "title": "Task2",
  },
]
`;

exports[`TodoListComponent should filter out items with empty title: todoItems 1`] = `
[
  {
    "title": "Task1",
  },
]
`;

exports[`TodoListComponent should navigate to details page 1`] = `
[
  [
    "ITEM_ID_1",
  ],
  {
    "relativeTo": "ACTIVATED_ROUTE",
  },
]
`;

exports[`TodoListComponent template should render todo list item 1`] = `
[
  [
    {
      "completed": false,
      "title": "Task1",
    },
  ],
]
`;
